crates/eips/src/eip7594/sidecar.rs

```diff
@@ -1,17 +1,132 @@
+/// Proof type depends on the sidecar variant.
+///
+/// This type encodes and decodes the fields without an rlp header.
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "serde", serde(untagged))]
+#[cfg_attr(any(test, feature = "arbitrary"), derive(arbitrary::Arbitrary))]
+pub enum BlobTransactionSidecarVariant {
+    /// EIP-4844 style blob transaction sidecar.
+    Eip4844(BlobTransactionSidecar),
+    /// EIP-7594 style blob transaction sidecar with cell proofs.
+    Eip7594(BlobTransactionSidecarEip7594),
+}
+
+impl BlobTransactionSidecarVariant {
+    /// Calculates a size heuristic for the in-memory size of the [BlobTransactionSidecarVariant].
+    #[inline]
+    pub fn size(&self) -> usize {
+        match self {
+            Self::Eip4844(sidecar) => sidecar.size(),
+            Self::Eip7594(sidecar) => sidecar.size(),
+        }
+    }
+
+    /// Verifies that the sidecar is valid. See relevant methods for each variant for more info.
+    #[cfg(feature = "kzg")]
+    pub fn validate(
+        &self,
+        blob_versioned_hashes: &[B256],
+        proof_settings: &c_kzg::KzgSettings,
+    ) -> Result<(), BlobTransactionValidationError> {
+        match self {
+            Self::Eip4844(sidecar) => sidecar.validate(blob_versioned_hashes, proof_settings),
+            Self::Eip7594(sidecar) => sidecar.validate(blob_versioned_hashes, proof_settings),
+        }
+    }
+
+    /// Returns an iterator over the versioned hashes of the commitments.
+    pub fn versioned_hashes(&self) -> Vec<B256> {
+        match self {
+            Self::Eip4844(sidecar) => sidecar.versioned_hashes().collect(),
+            Self::Eip7594(sidecar) => sidecar.versioned_hashes().collect(),
+        }
+    }
+
+    /// Outputs the RLP length of the [BlobTransactionSidecarVariant] fields, without a RLP header.
+    #[doc(hidden)]
+    pub fn rlp_encoded_fields_length(&self) -> usize {
+        match self {
+            Self::Eip4844(sidecar) => sidecar.rlp_encoded_fields_length(),
+            Self::Eip7594(sidecar) => sidecar.rlp_encoded_fields_length(),
+        }
+    }
+
+    /// Encodes the inner [BlobTransactionSidecarVariant] fields as RLP bytes, __without__ a RLP
+    /// header.
+    #[inline]
+    #[doc(hidden)]
+    pub fn rlp_encode_fields(&self, out: &mut dyn BufMut) {
+        match self {
+            Self::Eip4844(sidecar) => sidecar.rlp_encode_fields(out),
+            Self::Eip7594(sidecar) => sidecar.rlp_encode_fields(out),
+        }
+    }
+
+    /// RLP decode the fields of a [BlobTransactionSidecarVariant] based on the wrapper version.
+    #[doc(hidden)]
+    pub fn rlp_decode_fields(buf: &mut &[u8]) -> alloy_rlp::Result<Self> {
+        if buf.first() == Some(&EIP_7594_WRAPPER_VERSION) {
+            buf.advance(1);
+            Ok(Self::Eip7594(BlobTransactionSidecarEip7594::rlp_decode_fields(buf)?))
```

Comment:
this feels a bit strange, because this kinda behaves like a type tag

that probably shouldn't be encoded like this:

https://github.com/alloy-rs/alloy/blob/cce24039fa259559453d217eb55d7d7eae28d328/crates/eips/src/eip7594/sidecar.rs#L207-L208

because this isn't symmetrical

so one way is wrong I believe

could we add a simple roundtrip test for this as well?